streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1858,12 +1909,22 @@ object ZStream extends ZStreamPlatformSpecific {
 
       override def process: ZManaged[R, E, InputStream[R, E, A]] =
         for {
-          done <- Ref.make(false).toManaged_
-          a    <- managed
-        } yield done.get.flatMap {
-          if (_) InputStream.end
-          else done.set(true) *> InputStream.emit(a)
-        }
+          doneRef      <- Ref.make(false).toManaged_
+          finalizerRef <- Ref.make[Exit[_, _] => ZIO[R, Nothing, Any]](_ => UIO.unit).toManaged_
+          _            <- ZManaged.finalizerExit(e => finalizerRef.get.flatMap(_.apply(e)))
+          pull = ZIO.uninterruptibleMask { restore =>
```

Comment:
If we had a helper:

```scala
object InputStream {
  def singleton(a: A): UIO[InputStream[Any, Nothing, A]] = 
    Ref.make[Option[A]](Some(a)).map(ref =>
      ref.modify {
        case None => ZIO.fail(()) -> None
        case Some(a) => ZIO.succeed(a) -> None
      }.flatten)
```

Then it might make this one easier to verify.

Thinking what you want to do is map the `A` inside the managed into the input stream. The current version seems quite aggressive about uninterruptibility...

Reply:
That helper looks useful!

The logic for this constructor is almost what you wrote; I want the managed to allocate the `A` only when the InputStream is pulled. Does that make sense?

Reply:
It does, but I can't convince myself there's a difference because of the contract on the original `ZManaged[..., A]` that is being lifted.

Reply:
I *think* (not sure yet!) it matters when composing streams together. You're lifting a managed resource to a single-element stream and then flatMap it to create a multiple-element stream. So it should only allocate when you pull the composed stream.